### PR TITLE
Implementation of Queue Using Single Stack

### PR DESCRIPTION
**Pull request by Mysterious786**
#hacktoberfest 2k22

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
Implementation of Queue using two stacks .

### CODE REVIEW

**Review by Mysterious786 — COMMENTED** [review]
changed file name

**Comment by Mysterious786** [comment]
Do let me know if it needs to be correct or if it not valid.

**Comment by Mysterious786** [comment]
please

File: data_structures/Queue_implementation.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// CPP program to implement Queue using
+// two stacks with costly deQueue()
+#include <bits/stdc++.h>
+using namespace std;
```
Comment (Panquesito7) [inline data_structures/Queue_implementation.cpp:4]: These are discouraged. Please use each header/library you need and strictly use `std::`.

File: data_structures/Queue_implementation.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+struct Queue {
+	stack<int> s1, s2;
+
+	// Enqueue an item to the queue
+	void enQueue(int x)
```
Comment (Panquesito7) [inline data_structures/Queue_implementation.cpp:10]: Functions should be documented explaining the use of its parameters, a brief description, and what it returns.

File: data_structures/Queue_implementation.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
+// CPP program to implement Queue using
+// two stacks with costly deQueue()
+#include <bits/stdc++.h>
+using namespace std;
+
+struct Queue {
+	stack<int> s1, s2;
```
Comment (Panquesito7) [inline data_structures/Queue_implementation.cpp:7]: Please add a one-line description of what the variable is for (**see the example below**).

```c++
int64_t data = 0;       ///< The value/key of the node.
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/Queue_implementation.cpp (hunk 3 above)
Comment (Mysterious786) [reply data_structures/Queue_implementation.cpp:7]: ok Thank you for you suggestion , will correct it soon

File: data_structures/Queue_implementation.cpp (hunk 1 above)
Comment (Mysterious786) [reply data_structures/Queue_implementation.cpp:4]: ok

**Comment by Mysterious786** [comment]
Updated the code with comments check it please.

**Comment by Mysterious786** [comment]
### Output of the updated code


Screenshot (143)

**Comment by Panquesito7** [comment]
> Updated the code with comments check it please.

Still missing proper documentation. You can check the [**typical structure of a program**](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) if you're confused about what changes to make.

**Comment by Mysterious786** [comment]
ok , I will try to make it according to the structure. Thank you for your help

**Comment by Mysterious786** [comment]
Tried to make it good

**Comment by Mysterious786** [comment]
Can you please check I tried to improve documentation

**Comment by Mysterious786** [comment]
While reading different documentation and codes I found here that Queue Implementation using double Stacks is already present so that's why I Changed my code to Queue Implementation using Single Linked List.

**Comment by Mysterious786** [comment]
Can you please recheck it.

File: data_structures/queue_implementation.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
+/**
+ * @author [saqlain] (https://github.com/Mysterious786)
+ * @file
+ *
+ * Implementation of a Queue using Single Stacks.
```
Comment (Panquesito7) [inline data_structures/queue_implementation.cpp:5]: Please describe how the algorithm works, a Wikipedia or any other source for algorithm explanation, what the algorithm does, etc..

File: data_structures/queue_implementation.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ */
+
+#include <iostream>
+#include <stack>
+#include <cstdlib>
```
Comment (Panquesito7) [inline data_structures/queue_implementation.cpp:11]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Good work! 🚀

File: data_structures/queue_implementation.cpp (hunk 5 above)
Comment (Mysterious786) [reply data_structures/queue_implementation.cpp:11]: Ok

File: data_structures/queue_implementation.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
+/**
+ * @author [saqlain] (https://github.com/Mysterious786)
```
Comment (Panquesito7) [inline data_structures/queue_implementation.cpp:2]:
```suggestion
 * @author [saqlain](https://github.com/Mysterious786)
```

File: data_structures/queue_implementation.cpp (hunk 4 above)
Comment (Panquesito7) [inline data_structures/queue_implementation.cpp:5]: Please describe how the algorithm works, a Wikipedia or any other source for algorithm explanation, what the algorithm does, etc..

File: data_structures/queue_implementation.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+ * @file
+ *
+ * Implementation of a Queue using Single Stacks.
+ *
+ */
+
+#include <iostream>  /// for IO operations
+#include <stack>     /// for stack operations
+#include <cstdlib>   /// for datatype conversion
+namespace
```
Comment (Panquesito7) [inline data_structures/queue_implementation.cpp:12]:
```suggestion
/**
 * @brief Data structures
 * @namespace
 */
namespace data_structures
```

File: data_structures/queue_implementation.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+    class Queue
+    {
+        std::stack<int> s;
+
+    public:
+        /**
+         * Pushes data to the back of queue
+         */
+        void enqueue(int data)
```
Comment (Panquesito7) [inline data_structures/queue_implementation.cpp:28]: Parameters should be documented as well.

File: data_structures/queue_implementation.cpp [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            s.push(top);
+
+            
+            /**
+             * return the result of dequeue() call
+             */
+            return item;
+        }
+    };
+}
```
Comment (Panquesito7) [inline data_structures/queue_implementation.cpp:82]: Please add self-test cases so we can verify your program works properly (**see the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) as a reference**). Let us know if you need any help here or on our [**Discord server**](https://the-algorithms.com/discord). 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Good work! 👍

File: data_structures/queue_implementation.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+#include <stack>     /// for stack operations
+#include <cstdlib>   /// for datatype conversion
+/**
+ * @brief Data structures
+ * @namespace
+ */
+namespace data_structures
+{
+    /**
+     * Queue data structure.Store elements in FIFO
```
Comment (krishnacx) [inline data_structures/queue_implementation.cpp:19]: . Store

File: data_structures/queue_implementation.cpp (hunk 8 above)
Comment (krishnacx) [reply data_structures/queue_implementation.cpp:32]: Can we rename the argument to item than data ?

File: data_structures/queue_implementation.cpp [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+             * Determines whether the stack is empty
+             */
+            if (s.empty())
+            {
+                std::cout << "Underflow!!";
+                exit(0);
+            }
+
+            /**
+             *Removes item from the Stack
```
Comment (krishnacx) [inline data_structures/queue_implementation.cpp:55]: * Removes

File: data_structures/queue_implementation.cpp [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+             */
+            int top = s.top();
+            s.pop();
+
+            
+            /**
+             * Determines whether the stack becomes empty,and returns the popped item
```
Comment (krishnacx) [inline data_structures/queue_implementation.cpp:62]: empty and returns

File: data_structures/queue_implementation.cpp [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+             * Recursion call
+             */
+            int item = dequeue();
+
+            /**
+             * push popped item back into the stack
+             */
+            s.push(top);
+
+
```
Comment (krishnacx) [inline data_structures/queue_implementation.cpp:79]: Multiple empty line

File: data_structures/queue_implementation.cpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    for (int key : keys)
+    {
+        q.enqueue(key);
+    }
+    /**
+     * Print
+     *
+     */
+    std::cout << q.dequeue() << " ";
+    std::cout << q.dequeue() << " ";
```
Comment (krishnacx) [inline data_structures/queue_implementation.cpp:109]: Why multiple invocations ?

**Pull request closed without merging**